src/contrib/talloc.rst

```diff
@@ -0,0 +1,394 @@
+        {
+            free(sysdb->file);
+            free_ldb(sysdb->ldb);
+            free(ldb);
+        }
+
+        void free_domain(struct sss_domain_info *domain)
+        {
+            free(name);
+            free(provider);
+            free_sysdb(domain->sysdb);
+            free_names(domain->names)
+            free(domain);
+        }
+
+        free_domain(domain);
+
+.. _confdb.h: https://github.com/SSSD/sssd/blob/master/src/confdb/confdb.h#L353
+
+.. note::
+
+    Every pointer created by talloc is a memory context on its own.
+
+Changing the hierarchy
+======================
+
+Ideally, what you want to create is a nice oriented graph where nodes are
+structures and leaves are non-structure elements (e.g. strings). The path must
+follow the natural ordering that comes from the structure definition -- that is
+structures are the parent memory contexts and their fields are their children as
+you can see from previous examples.
+
+Sometimes (e.g. when your function produces an output values) you want to change
+the parent context. You can use ``talloc_steal`` or ``talloc_move`` for this
+operation.
+
+.. code-block:: c
+
+    char *value = talloc_strdup(NULL, "Hello world!");
+    struct record *r;
+
+    r = talloc_zero(NULL, struct record);
+    r->name = talloc_strdup(r, "Hello");
+    r->value = talloc_steal(r, value)
+
+.. seealso::
+
+    There are many functions that can create a new talloc memory context. These
+    functions always take an existing talloc context as an argument that is used
+    as the parent context. For example ``talloc_new``, ``talloc_zero``,
+    ``talloc_zero_array``, ``talloc_strdup``, ``talloc_strndup`` and
+    ``talloc_asprintf``.
+
+Destructors
+***********
+
+It is possible to assign a destructor on a memory context. The destructor is a
+function that takes the context as an argument and is executed when the context
+is freed before any of its children are freed. This can be used to perform
+additional clean up when simple memory free is not sufficient.
+
+.. code-block:: c
+
+    int record_destructor(struct record *r)
+    {
+        printf("Removing %s\n", r->name);
+
+        return 0;
+    }
+
+
+    r = talloc_zero(NULL, struct record);
+    r->name = talloc_strdup(r, "Hello");
+    r->value = talloc_strdup(r, "World");
+
+    talloc_set_destructor(r, record_destructor);
+
+.. seealso::
+
+    Checkout the example from the official `Tutorial
+    <https://talloc.samba.org/talloc/doc/html/libtalloc__destructors.html>`_.
+
+Identifying memory leaks
+************************
+
+Talloc has the ability to produce dump of a complete tree from the first top
+level context down to the leaves. This information is very useful for debugging
+a memory leak since it can help you identify the talloc context that was not
+freed properly. Use the following snippet to produce the talloc report from a
+running process:
+
+.. code-block:: bash
+
+    PROCESS=$(pidof sssd_nss)
```

Comment:
I would add information that talloc dump can be called from source code too.
Also there is a part of the documentation maybe worth mentioning:
```
Apart from the abort function, talloc uses a log function to provide additional information to the aforementioned violations. To enable logging we shall set the log function with one of:

    talloc_set_log_fn()
    talloc_set_log_stderr()

```

Reply:
> I would add information that talloc dump can be called from source code too.

It is not something we use, therefore its not important for  us.

> Also there is a part of the documentation maybe worth mentioning:

This is covered in the Tutorial. It's already enabled in SSSD so it's not really important for new contributors to get started.

Note that this is a getting started guide, not a  comprehensive talloc documentation. For that there is the Tutorial and API reference.